File: Medium/Java/Find_Peak_Element(162).java
class Solution {
    public int findPeakElement(int[] nums) {
        int l = 0;
        int r = nums.length - 1;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if ((mid == 0 || nums[mid - 1] < nums[mid]) && (mid == nums.length - 1 || nums[mid + 1] < nums[mid]))
                return mid;
            if (mid > 0 && nums[mid] < nums[mid - 1]) 
                r = mid - 1;
            else 
                l = mid + 1;
        }
        return -1;
    }
}

File: Medium/C++/Find_Peak_Element(162).cpp
class Solution {
public:
    //use concept of finding element in segment tree (getval function)
    //or concept similar to merge sort
    int ind=-1;
    int getVal(vector<int>& nums,int i,int j){
        if(i==j)return i;
        int m=(i+j)/2;
        int l=getVal(nums,i,m);
        int r=getVal(nums,m+1,j);
        return nums[l]>=nums[r]?l:r;
    }
    int findPeakElement(vector<int>& nums) {
        return getVal(nums,0,nums.size()-1);
    }
};